File: hread.go
package hpatch

import (
	"context"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"regexp"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/yusing/hpatch/internal/hpatchsyntax"
)

const (
	maxHReadOutputBytes    = 16 << 20
	maxHReadBatchItems     = 32
	hreadBufferBytes       = 32 << 10
	hreadBatchLimitMessage = "hread: batch output limit reached; retry remaining items in a narrower batch\n"
)

var (
	hreadRangePattern = regexp.MustCompile(`^ ([1-9][0-9]*):([1-9][0-9]*)$`)

	// ErrHReadResultTooLarge reports that hashline formatting would exceed the
	// bounded result size enforced by hread and its router integration.
	ErrHReadResultTooLarge = errors.New("hread result exceeds its configured bound")
)

// HashLineReadResult is the host-facing result of one hread call.
type HashLineReadResult struct {
	Output string
}

// ReadHashLines reads one grammar-constrained hread call through workspace and
// returns verified LINE:HASH logical rows for one or more file specifications.
func ReadHashLines(ctx context.Context, workspace Workspace, input string) (string, error) {
	result, err := ReadHashLinesForHost(ctx, workspace, input)
	return result.Output, err
}

func parseHReadSpec(input string) (path string, startLine, endLine int, err error) {
	path, trailing, err := hpatchsyntax.DecodeQuoted(input)
	if err != nil {
		return "", 0, 0, fmt.Errorf("invalid hread path: %w", err)
	}
	if path == "" {
		return "", 0, 0, errors.New("hread path must not be empty")
	}
	if trailing == "" {
		return path, 0, 0, nil
	}
	match := hreadRangePattern.FindStringSubmatch(trailing)
	if match == nil {
		return "", 0, 0, errors.New(`hread input must be "PATH" or "PATH" START:END`)
	}
	startLine, err = strconv.Atoi(match[1])
	if err != nil {
		return "", 0, 0, errors.New("hread start line is out of range")
	}
	endLine, err = strconv.Atoi(match[2])
	if err != nil {
		return "", 0, 0, errors.New("hread end line is out of range")
	}
	if startLine > endLine {
		return "", 0, 0, errors.New("hread line range start exceeds end")
	}
	return path, startLine, endLine, nil
}

// ReadHashLinesForHost returns the formatted hread output for host integrations.
func ReadHashLinesForHost(ctx context.Context, workspace Workspace, input string) (HashLineReadResult, error) {
	return readHashLinesForHost(ctx, workspace, input, maxHReadOutputBytes)
}

func readHashLinesForHost(ctx context.Context, workspace Workspace, input string, maxOutputBytes int) (HashLineReadResult, error) {
	rawSpecs := strings.Split(input, "\n")
	if len(rawSpecs) > maxHReadBatchItems {
		return HashLineReadResult{}, fmt.Errorf("hread batch exceeds %d items", maxHReadBatchItems)
	}

	type readSpec struct {
		input              string
		path               string
		startLine, endLine int
		err                error
	}
	specs := make([]readSpec, len(rawSpecs))
	for index, raw := range rawSpecs {
		raw = strings.TrimSuffix(raw, "\r")
		if raw == "" {
			return HashLineReadResult{}, errors.New("hread batch contains an empty read specification")
		}
		path, startLine, endLine, err := parseHReadSpec(raw)
		specs[index] = readSpec{
			input:     raw,
			path:      path,
			startLine: startLine,
			endLine:   endLine,
			err:       err,
		}
	}
	if len(specs) == 1 && specs[0].err != nil {
		return HashLineReadResult{}, specs[0].err
	}

	filesystem, err := validateWorkspace(ctx, workspace)
	if err != nil {
		return HashLineReadResult{}, err
	}
	read := func(spec readSpec, limit int) (HashLineReadResult, error) {
		if spec.err != nil {
			return HashLineReadResult{}, spec.err
		}
		path, err := filesystem.resolvePath(spec.path)
		if err != nil {
			return HashLineReadResult{}, err
		}
		return filesystem.readHashLines(ctx, path, spec.startLine, spec.endLine, limit)
	}
	if len(specs) == 1 {
		return read(specs[0], maxOutputBytes)
	}

	dataLimit := max(0, maxOutputBytes-len(hreadBatchLimitMessage))
	var output strings.Builder
	appendBounded := func(text string) bool {
		if len(text) > dataLimit-output.Len() {
			return false
		}
		output.WriteString(text)
		return true
	}
	appendLimitMessage := func() {
		if len(hreadBatchLimitMessage) <= maxOutputBytes-output.Len() {
			output.WriteString(hreadBatchLimitMessage)
		}
	}
	for _, spec := range specs {
		if err := ctx.Err(); err != nil {
			return HashLineReadResult{}, err
		}
		if !appendBounded(fmt.Sprintf("==> %s <==\n", spec.input)) {
			appendLimitMessage()
			break
		}
		result, err := read(spec, dataLimit-output.Len())
		if contextErr := ctx.Err(); contextErr != nil {
			return HashLineReadResult{}, contextErr
		}
		if errors.Is(err, ErrHReadResultTooLarge) {
			appendLimitMessage()
			break
		}
		if err != nil {
			if !appendBounded(fmt.Sprintf("hread: %s\n", err)) {
				appendLimitMessage()
				break
			}
			continue
		}
		output.WriteString(result.Output)
	}
	return HashLineReadResult{Output: output.String()}, nil
}

func (w filesystemWorkspace) openRegularFile(ctx context.Context, path string) (*os.File, fs.FileMode, error) {
	if err := ctx.Err(); err != nil {
		return nil, 0, err
	}
	file, err := w.root.Open(path)
	if err != nil {
		reason := reasonOther
		if errors.Is(err, fs.ErrNotExist) {
			reason = reasonFileMissing
		}
		return nil, 0, withReason(reason, fmt.Errorf("reading %s: %w", path, err))
	}
	info, err := file.Stat()
	if err != nil {
		file.Close()
		return nil, 0, fmt.Errorf("reading %s: %w", path, err)
	}
	if !info.Mode().IsRegular() {
		file.Close()
		return nil, 0, fmt.Errorf("%s is not a regular file", path)
	}
	return file, info.Mode(), nil
}

func (w filesystemWorkspace) readFile(ctx context.Context, path string) (loadedFile, error) {
	file, mode, err := w.openRegularFile(ctx, path)
	if err != nil {
		return loadedFile{}, err
	}
	defer file.Close()

	content, err := io.ReadAll(file)
	if err != nil {
		return loadedFile{}, fmt.Errorf("reading %s: %w", path, err)
	}
	if !utf8.Valid(content) {
		return loadedFile{}, fmt.Errorf("%s is not UTF-8", path)
	}
	return loadedFile{content: string(content), mode: mode}, nil
}

func (w filesystemWorkspace) readHashLines(ctx context.Context, path string, startLine, endLine, maxOutputBytes int) (HashLineReadResult, error) {
	file, _, err := w.openRegularFile(ctx, path)
	if err != nil {
		return HashLineReadResult{}, err
	}
	defer file.Close()

	return formatHashLineStream(ctx, file, path, startLine, endLine, maxOutputBytes)
}

func formatHashLineStream(
	ctx context.Context,
	reader io.Reader,
	path string,
	startLine, endLine, maxOutputBytes int,
) (HashLineReadResult, error) {
	wholeFile := startLine == 0 && endLine == 0
	lineNumber := 1
	lineNumberWidth := 1
	lineOpen := false
	pendingCR := false

	var output strings.Builder
	var content strings.Builder
	selected := func() bool {
		return wholeFile || lineNumber >= startLine && lineNumber <= endLine
	}
	capacityError := func() error {
		return fmt.Errorf("%w of %d bytes", ErrHReadResultTooLarge, maxOutputBytes)
	}
	appendContent := func(character byte) error {
		lineOpen = true
		if !selected() {
			return nil
		}
		rowBytes := output.Len() + lineNumberWidth + 7 + content.Len() + 1
		if rowBytes > maxOutputBytes {
			return capacityError()
		}
		content.WriteByte(character)
		return nil
	}
	finishLine := func() error {
		if selected() {
			rowBytes := output.Len() + lineNumberWidth + 7 + content.Len()
			if rowBytes > maxOutputBytes {
				return capacityError()
			}
			lineContent := content.String()
			writeHashLine(&output, lineNumber, lineContent, lineContent)
		}
		content.Reset()
		lineNumber++
		lineNumberWidth = len(strconv.Itoa(lineNumber))
		lineOpen = false
		return nil
	}

	var encodedRune [utf8.UTFMax]byte
	encodedRuneBytes := 0
	validateByte := func(character byte) bool {
		encodedRune[encodedRuneBytes] = character
		encodedRuneBytes++
		encoded := encodedRune[:encodedRuneBytes]
		if !utf8.FullRune(encoded) {
			return true
		}
		decoded, size := utf8.DecodeRune(encoded)
		if decoded == utf8.RuneError && size == 1 {
			return false
		}
		encodedRuneBytes = 0
		return true
	}

	buffer := make([]byte, hreadBufferBytes)
	emptyReads := 0
	for {
		if err := ctx.Err(); err != nil {
			return HashLineReadResult{}, err
		}
		readBytes, readErr := reader.Read(buffer)
		if err := ctx.Err(); err != nil {
			return HashLineReadResult{}, err
		}
		if readBytes == 0 && readErr == nil {
			emptyReads++
			if emptyReads >= 100 {
				return HashLineReadResult{}, fmt.Errorf("reading %s: %w", path, io.ErrNoProgress)
			}
			continue
		}
		emptyReads = 0

		for _, character := range buffer[:readBytes] {
			if !validateByte(character) {
				return HashLineReadResult{}, fmt.Errorf("%s is not UTF-8", path)
			}
			if pendingCR {
				pendingCR = false
				if character == '\n' {
					if err := finishLine(); err != nil {
						return HashLineReadResult{}, err
					}
					continue
				}
				if err := finishLine(); err != nil {
					return HashLineReadResult{}, err
				}
			}
			switch character {
			case '\r':
				lineOpen = true
				pendingCR = true
			case '\n':
				lineOpen = true
				if err := finishLine(); err != nil {
					return HashLineReadResult{}, err
				}
			default:
				if err := appendContent(character); err != nil {
					return HashLineReadResult{}, err
				}
			}
		}
		if readErr != nil {
			if !errors.Is(readErr, io.EOF) {
				return HashLineReadResult{}, fmt.Errorf("reading %s: %w", path, readErr)
			}
			break
		}
	}
	if encodedRuneBytes != 0 {
		return HashLineReadResult{}, fmt.Errorf("%s is not UTF-8", path)
	}
	if pendingCR {
		if err := finishLine(); err != nil {
			return HashLineReadResult{}, err
		}
	} else if lineOpen {
		if err := finishLine(); err != nil {
			return HashLineReadResult{}, err
		}
	}
	if err := ctx.Err(); err != nil {
		return HashLineReadResult{}, err
	}

	lineCount := lineNumber - 1
	if !wholeFile && (startLine < 1 || endLine < startLine || startLine > lineCount) {
		return HashLineReadResult{}, fmt.Errorf(
			"requested lines %d:%d are outside file with %d lines",
			startLine,
			endLine,
			lineCount,
		)
	}
	return HashLineReadResult{Output: output.String()}, nil
}

File: internal/router/toolplugin/src/builtin/hread.ts
import {constants} from "node:fs";
import {open} from "node:fs/promises";

import type {Tool} from "../../plugin.d.ts";

import {
  MAX_OUTPUT_BYTES,
  byteLength,
  errorText,
  createExecutorTool,
  formatHashLine,
} from "./common.ts";

// Source: hread.go:45:385 parseHReadSpec, readHashLinesForHost, and formatHashLineStream.
const MAX_BATCH_ITEMS = 6;
const READ_BUFFER_BYTES = 32 * 1024;
const BATCH_LIMIT_MESSAGE = "hread: batch output limit reached; retry remaining items in a narrower batch\n";

class ResultTooLargeError extends Error {}

type ReadSpec = {
  input: string;
  path: string;
  startLine: number;
  endLine: number;
};

type ParsedReadSpec = {
  spec: ReadSpec;
  error: unknown | null;
};

function parseQuotedPath(input: string): {path: string; trailing: string} {
  let escaped = false;
  for (let index = 1; index < input.length; index += 1) {
    const character = input[index];
    if (escaped) {
      escaped = false;
      continue;
    }
    if (character === "\\") {
      escaped = true;
      continue;
    }
    if (character !== "\"") {
      continue;
    }
    const encoded = input.slice(0, index + 1);
    let path;
    try {
      path = JSON.parse(encoded);
    } catch (error) {
      throw new Error(`invalid hread path: ${errorText(error)}`);
    }
    if (typeof path !== "string") {
      throw new Error("invalid hread path");
    }
    return {path, trailing: input.slice(index + 1)};
  }
  throw new Error("invalid hread path: unterminated quoted string");
}

function parseReadSpec(input: string): ReadSpec {
  let path;
  let trailing;
  if (input.startsWith("\"")) {
    ({path, trailing} = parseQuotedPath(input));
  } else {
    const separator = input.indexOf(" ");
    path = separator < 0 ? input : input.slice(0, separator);
    trailing = separator < 0 ? "" : input.slice(separator);
    if (/[\u0000-\u0020"]/u.test(path)) {
      throw new Error("invalid bare hread path");
    }
  }
  if (path === "") {
    throw new Error("hread path must not be empty");
  }
  if (trailing === "") {
    return {input, path, startLine: 0, endLine: 0};
  }
  const match = trailing.match(/^ ([1-9][0-9]*):([1-9][0-9]*)$/u);
  if (match === null) {
    throw new Error("hread input must be PATH or PATH START:END");
  }
  const startLine = Number(match[1]);
  const endLine = Number(match[2]);
  if (!Number.isSafeInteger(startLine)) {
    throw new Error("hread start line is out of range");
  }
  if (!Number.isSafeInteger(endLine)) {
    throw new Error("hread end line is out of range");
  }
  if (startLine > endLine) {
    throw new Error("hread line range start exceeds end");
  }
  return {input, path, startLine, endLine};
}

function parseReadSpecs(input: string): ParsedReadSpec[] {
  const rawSpecs = input.split("\n");
  if (rawSpecs.length > MAX_BATCH_ITEMS) {
    throw new Error(`hread batch exceeds ${MAX_BATCH_ITEMS} items`);
  }
  const specs = rawSpecs.map((raw) => {
    const normalized = raw.endsWith("\r") ? raw.slice(0, -1) : raw;
    if (normalized === "") {
      throw new Error("hread batch contains an empty read specification");
    }
    try {
      return {spec: parseReadSpec(normalized), error: null};
    } catch (error) {
      return {
        spec: {input: normalized, path: "", startLine: 0, endLine: 0},
        error,
      };
    }
  });
  if (specs.length === 1 && specs[0].error !== null) {
    throw specs[0].error;
  }
  return specs;
}

async function readHashLines(spec: ReadSpec, maxOutputBytes: number): Promise<string> {
  let handle;
  try {
    handle = await open(spec.path, constants.O_RDONLY | (constants.O_NONBLOCK ?? 0));
  } catch (error) {
    throw new Error(`reading ${spec.path}: ${errorText(error)}`);
  }

  try {
    const info = await handle.stat();
    if (!info.isFile()) {
      throw new Error(`${spec.path} is not a regular file`);
    }

    const wholeFile = spec.startLine === 0 && spec.endLine === 0;
    let lineNumber = 1;
    let lineOpen = false;
    let pendingCR = false;
    let content = "";
    let contentBytes = 0;
    let output = "";
    let outputBytes = 0;

    const selected = () => wholeFile
      || (lineNumber >= spec.startLine && lineNumber <= spec.endLine);
    const capacityError = () => new ResultTooLargeError(
      `hread result exceeds its configured bound of ${maxOutputBytes} bytes`,
    );
    const appendContent = (text: string): void => {
      if (text === "") {
        return;
      }
      lineOpen = true;
      if (!selected()) {
        return;
      }
      const addedBytes = byteLength(text);
      const rowFramingBytes = byteLength(String(lineNumber)) + 7;
      if (outputBytes + rowFramingBytes + contentBytes + addedBytes > maxOutputBytes) {
        throw capacityError();
      }
      content += text;
      contentBytes += addedBytes;
    };
    const finishLine = (): void => {
      if (selected()) {
        const row = formatHashLine(lineNumber, content);
        const rowBytes = byteLength(row);
        if (outputBytes + rowBytes > maxOutputBytes) {
          throw capacityError();
        }
        output += row;
        outputBytes += rowBytes;
      }
      content = "";
      contentBytes = 0;
      lineNumber += 1;
      lineOpen = false;
    };
    const consume = (text: string): void => {
      let offset = 0;
      while (offset < text.length) {
        if (pendingCR) {
          pendingCR = false;
          finishLine();
          if (text[offset] === "\n") {
            offset += 1;
            continue;
          }
        }
        const cr = text.indexOf("\r", offset);
        const lf = text.indexOf("\n", offset);
        let end = text.length;
        if (cr >= 0) {
          end = cr;
        }
        if (lf >= 0 && lf < end) {
          end = lf;
        }
        appendContent(text.slice(offset, end));
        if (end === text.length) {
          break;
        }
        lineOpen = true;
        if (text[end] === "\r") {
          pendingCR = true;
        } else {
          finishLine();
        }
        offset = end + 1;
      }
    };

    const decoder = new TextDecoder("utf-8", {fatal: true});
    const stream = handle.createReadStream({
      autoClose: false,
      highWaterMark: READ_BUFFER_BYTES,
    });
    try {
      for await (const chunk of stream) {
        consume(decoder.decode(chunk, {stream: true}));
      }
      consume(decoder.decode());
    } catch (error) {
      if (error instanceof ResultTooLargeError) {
        throw error;
      }
      if (error instanceof TypeError) {
        throw new Error(`${spec.path} is not UTF-8`);
      }
      throw new Error(`reading ${spec.path}: ${errorText(error)}`);
    } finally {
      stream.destroy();
    }

    if (pendingCR) {
      finishLine();
    } else if (lineOpen) {
      finishLine();
    }
    const lineCount = lineNumber - 1;
    if (!wholeFile && spec.startLine > lineCount) {
      throw new Error(
        `requested lines ${spec.startLine}:${spec.endLine} are outside file with ${lineCount} lines`,
      );
    }
    return output;
  } finally {
    await handle.close();
  }
}

async function executeRead(input: string): Promise<string> {
  const specs = parseReadSpecs(input);
  if (specs.length === 1) {
    return readHashLines(specs[0].spec, MAX_OUTPUT_BYTES);
  }

  const dataLimit = Math.max(0, MAX_OUTPUT_BYTES - byteLength(BATCH_LIMIT_MESSAGE));
  let output = "";
  let outputBytes = 0;
  const appendBounded = (text: string): boolean => {
    const addedBytes = byteLength(text);
    if (outputBytes + addedBytes > dataLimit) {
      return false;
    }
    output += text;
    outputBytes += addedBytes;
    return true;
  };
  const appendLimitMessage = (): void => {
    if (outputBytes + byteLength(BATCH_LIMIT_MESSAGE) <= MAX_OUTPUT_BYTES) {
      output += BATCH_LIMIT_MESSAGE;
      outputBytes += byteLength(BATCH_LIMIT_MESSAGE);
    }
  };

  for (const item of specs) {
    if (!appendBounded(`==> ${item.spec.input} <==\n`)) {
      appendLimitMessage();
      break;
    }
    if (item.error !== null) {
      if (!appendBounded(`hread: ${errorText(item.error)}\n`)) {
        appendLimitMessage();
        break;
      }
      continue;
    }
    try {
      const result = await readHashLines(item.spec, dataLimit - outputBytes);
      output += result;
      outputBytes += byteLength(result);
    } catch (error) {
      if (error instanceof ResultTooLargeError) {
        appendLimitMessage();
        break;
      }
      if (!appendBounded(`hread: ${errorText(error)}\n`)) {
        appendLimitMessage();
        break;
      }
    }
  }
  return output;
}

export function createHReadTool(description: string, grammar: string): Tool<string> {
  return createExecutorTool({
    name: "hread",
    description,
    grammar,
    argv(input) {
      return [input];
    },
    async execute(argv) {
      if (argv.length !== 1) {
        return {
          stderr: "hread: expected one complete grammar input argument\n",
          exitCode: 1,
        };
      }
      try {
        return {stdout: await executeRead(argv[0]), exitCode: 0};
      } catch (error) {
        return {stderr: `hread: ${errorText(error)}\n`, exitCode: 1};
      }
    },
  });
}
